src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala

```diff
@@ -438,27 +414,19 @@ class SchedulerActions(
     instancesToKill.foreach { instances: Seq[Instance] =>
       logger.info(s"Scaling ${runSpec.id} from ${runningInstances.size} down to $targetCount instances")
 
-      launchQueue.purge(runSpec.id)
-        .recover {
-          case NonFatal(e) =>
-            logger.warn("Async purge failed: {}", e)
-            Done
-        }.foreach { _ =>
-          logger.info(s"Killing instances ${instances.map(_.instanceId)}")
-
-          def killInstances(instances: Seq[Instance]): Future[Done] = async {
-            val changeGoalsFuture = instances.map { i =>
-              if (i.hasReservation) instanceTracker.setGoal(i.instanceId, Goal.Stopped)
-              else instanceTracker.setGoal(i.instanceId, Goal.Decommissioned)
-            }
-
-            await(Future.sequence(changeGoalsFuture))
-            await(killService.killInstances(instances, KillReason.OverCapacity))
-          }
-
-          killInstances(instances)
-        }
+      val instancesAreTerminal = killService.watchForKilledInstances(instances)
+      val changeGoalsFuture = instances.map { i =>
+        if (i.hasReservation) instanceTracker.setGoal(i.instanceId, Goal.Stopped, GoalAdjustmentReason.OverCapacity)
+        else instanceTracker.setGoal(i.instanceId, Goal.Decommissioned, GoalAdjustmentReason.OverCapacity)
+      }
 
+      async {
+        await(launchQueue.purge(runSpec.id))
```

Comment:
I think there is a race. You start the goal change before you purge. Well, I hope that the purge is not required at all but right now the purge might override the goals again.

Reply:
hmm while thinking about this... can you maybe elaborate? I thought I found a race but then I actually realized that all this is happened while the app is locked but nothing else should actually mess with the goals while we do this...